### ISSUE TITLE
Render tag requests dynamically in the forum post

### ISSUE THREAD

**Issue by BrokenEagle** [r888888888/danbooru#3824]
When editing a tag alias, tag implication, or BUR, the text in the corresponding forum post does not get updated.  Many times the OP for that request remembers to update the forum post as well, but not always.  Additionally, this makes it difficult to update the request as anyone other than the OP or a Moderator+, as any change by a Builder+ does not automatically get reflected in the forum post.  Sometimes this can be rectified by Dmailing or @ mentioning the OP, but it shouldn't have to be reliant on such a mechanism.

Therefore, allow the content to be rendered dynamically given the actual information in the request, and just place it before any text in the actual forum post body.  This has the added benefit of simplifying the DText in the request itself.

Since this would affect old existing forum post requests, a new field might need to be added to forum posts to indicate if there is a request tied to that post and what type of request it is.  That way, the server will only need to look up the request information on only the forum posts that have that information set.  Additionally, it will grandfather in all of the old request forum posts, with all of the new forum posts going forward using the new system.

As a final note, it might be beneficial to add an ``updater_id`` field to all request types.  That way it will be known if someone besides the OP changes the request but forgets to leave a notice about in on the thread.

**Comment by BrokenEagle** [comment]
On a related note to the above, I discovered that there is a forum post that is no longer votable ([forum #149587](url)) because the user removed the the "->" in the text of the body, so the code no longer detects that the forum post is votable.

https://github.com/r888888888/danbooru/blob/308cfc2376272d3fd08cdd7a0398221e875d4426/app/models/forum_post.rb#L141-L143

If forum posts had the request type field as mentioned in the OP, then that could be used instead to detect if a forum posts is votable, instead of relying on a fragile scheme.

**Comment by chinatsu-chan** [comment]
Presently for particularly long BURs, updating the text of the forum post is a very tedious and error-prone, hence why in the above mentioned forum #149587, I just directed the viewer to open the request. It would be preferred to have the request dynamically embedded in the forum post, especially with request that take on several revisions before it is completed.

**Comment by evazion** [comment]
![image](https://user-images.githubusercontent.com/8430473/51001687-5f023000-14f6-11e9-8b38-b9f3c672b1ed.png)

Having to expand short requests just to see them is inconvenient. Requests shouldn't be rendered in an expand block, except perhaps if they're very long.

**Comment by BrokenEagle** [comment]
Yeah, I agree... that's kind of annoying. The text should be at least more than the screen height before putting it in an expand.  [forum #153311](https://danbooru.donmai.us/forum_posts/153311) seems to be around the size where everything should still be shown, whereas [forum #1234](https://danbooru.donmai.us/forum_posts/153457) should definitely be in an expand.  That first one has 45 lines, so we could call the "non-expand max" something even like 50.

**Comment by BrokenEagle** [comment]
Actually, quoting the BUR produces weird results which I can tell is going to cause no end of problems.

```
[quote]
Runty said:

[bur:1931]

Reason: All of the above tags are a subset of the Elementalist Lux skin, and thus should implicate the tag. The [[light_elementalist_lux]] tag already has the implication.
[/quote]
```

I don't know if that **[bur:1931]** needs to be as part of the forum body, but it probably shouldn't be in the quote result if it does.  Also, if it does need to be part of the forum body, what happens if the user edits their forum post and accidentally deletes it?

**Comment by evazion** [comment]
> I don't know if that **[bur:1931]** needs to be as part of the forum body, but it probably shouldn't be in the quote result if it does. Also, if it does need to be part of the forum body, what happens if the user edits their forum post and accidentally deletes it?

`[bur:1931]` is effectively dtext pseudo-syntax for embedding a bur. It gets replaced by the content of the BUR. So taking it out would mean the BUR is no longer be embedded in the post.

As far as quoting, I think it'd be more sensible to expand the `[bur:1931]` tag out to the content of the BUR when you quote the message. This way when you quote the OP your quote includes the BUR as it was at the time of your quote, not as it is currently. This would complicate things though.

**Comment by BrokenEagle** [comment]
I noticed when editing out the body of one of my recent BURs, that it doesn't have the DText psuedo-syntax as mentioned above.

https://danbooru.donmai.us/forum_posts/153948

However, it does when actually viewing it through the API.

https://danbooru.donmai.us/forum_posts/153948.xml

I then did a test on Testbooru where I created a BUR, opened the forum post for editing then saved it immediately with no changes.

http://testbooru.donmai.us/forum_posts/81

As can be seen, it no longer has the DText psuedo-syntax.

http://testbooru.donmai.us/forum_posts/81.xml

I then went to edit the BUR to something different, and as expected the BUR on the forum post did not update.

http://testbooru.donmai.us/bulk_update_requests/12

Also, for some reason, this seems to only affect forum posts that aren't the OP.

http://testbooru.donmai.us/forum_posts/80

If you go to edit the above, the DText psuedo-syntax will appear as expected.

**Comment by evazion** [comment]
Fixed in a118ec9.
